rustls/src/vecbuf.rs

```diff
@@ -99,6 +99,56 @@ impl ChunkVecBuffer {
         Ok(offs)
     }
 
+    #[cfg(feature = "read_buf")]
+    #[allow(unsafe_code)]
+    /// Read data out of this object, writing it into `buf`.
+    pub(crate) fn read_buf(&mut self, buf: &mut io::ReadBuf<'_>) -> io::Result<()> {
+        use std::mem::MaybeUninit;
+
+        while !self.is_empty() {
+            // There are three unsafe calls in this block to justify. First,
+            // `ReadBuf::unfilled_mut()` requires that we "not de-initialize portions of the buffer
+            // that have already been initialized." We write to the buffer using
+            // `std::ptr::copy_nonoverlapping`, with a `[u8]` slice from this `ChunkVecBuffer` as
+            // the source. All memory in the `[u8]` slice must be initialized, and the length of
+            // the copy is less than or equal to the length of the `[u8]` slice, so we will not
+            // de-initialize any of the `ReadBuf` buffer by performing this copy.
+            //
+            // Second, `std::ptr::copy_nonoverlapping` has several requirements. First, note that
+            // the source and destination buffers do not overlap because the source is part of
+            // `self`, which is mutably borrowed, so we can assume the other buffer does not alias
+            // it. As we are copying `T=u8`, the alignment requirements on the source and
+            // destination are trivially satisfied. Additionally, `u8` is `Copy`, so creating
+            // bitwise copies is acceptible. The count argument is less than or equal to both the
+            // length of the slice we are copying from, and the length of the unfilled part of the
+            // buffer we are copying to. Thus, the source pointer is valid for reads of this
+            // length, and the destination pointer is valid for writes of this length.
+            //
+            // Lastly, `ReadBuf::assume_init` requires that "the first `n` unfilled bytes of the
+            // buffer have already been initialized". We write into the unfilled part of the buffer
+            // using `std::ptr::copy_nonoverlapping`, and, as argued above, this causes these bytes
+            // to become initialized, since we are copying from another buffer that is initialized.
+            unsafe {
```

Comment:
So I haven't looked at this in detail, and I appreciate the thoroughness that is going into this and the commentary here, but my predominant question is why this doesn't call `ReadBuf::append()` with `&chunk[..used])`? It seems to me that is equivalently efficient and means we have a route to supporting this feature and maintaining `forbid(unsafe_code)`

Reply:
To expand: I'd really, really, really like to keep `forbid(unsafe_code)` crate-wide.

Reply:
Ah, thanks, I wasn't aware of that. Yeah, that's a safe abstraction of exactly what I was doing.